modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EElvis.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.painless.node;
+
+import org.elasticsearch.painless.AnalyzerCaster;
+import org.elasticsearch.painless.Definition;
+import org.elasticsearch.painless.Definition.Type;
+import org.elasticsearch.painless.Globals;
+import org.elasticsearch.painless.Locals;
+import org.elasticsearch.painless.Location;
+import org.elasticsearch.painless.MethodWriter;
+import org.objectweb.asm.Label;
+
+import java.util.Set;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * The Elvis operator ({@code ?:}), a null coalescing operator. Binary operator that evaluates the first expression and return it if it is
+ * non null. If the first expression is null then it evaluates the second expression and returns it.
+ */
+public class EElvis extends AExpression {
+    private AExpression lhs;
+    private AExpression rhs;
+    /**
+     * Should we unbox result of the lhs if it is non null? This is required when the result is being cast to a primitive
+     * ({@code int i = params.i ? : 0}). Unlink {@link EConditional} which can cast the left and right arms to the result, the Elvis
+     * operator has to handle when the lhs is null. So the best it can do is cast the lhs to a the boxed form of the primitive and then
+     * unbox it if it is non-null.
+     */
+    private boolean unboxLhs;
+
+    public EElvis(Location location, AExpression lhs, AExpression rhs) {
+        super(location);
+
+        this.lhs = requireNonNull(lhs);
+        this.rhs = requireNonNull(rhs);
+    }
+
+    @Override
+    void extractVariables(Set<String> variables) {
+        lhs.extractVariables(variables);
+        rhs.extractVariables(variables);
+    }
+
+    @Override
+    void analyze(Locals locals) {
+        lhs.expected = expected;
+        if (lhs.expected != null && lhs.expected.sort.primitive) {
+            lhs.expected = Definition.getType(lhs.expected.sort.boxed.getSimpleName());
```

Comment:
I'm a bit squeamish about this trick but it seems to work. I describe the problem on the javadoc for `unboxLhs`. I  keep thinking I'm missing something but I couldn't come up with failing code....